src/int/shift.rs

```diff
@@ -78,8 +78,8 @@ intrinsics! {
     #[avr_skip]
     #[maybe_use_optimized_c_shim]
     #[arm_aeabi_alias = __aeabi_llsl]
-    pub extern "C" fn __ashldi3(a: u64, b: u32) -> u64 {
-        a.ashl(b)
+    pub extern "C" fn __ashldi3(a: u64, b: core::ffi::c_uint) -> u64 {
+        a.ashl(b as u32)
```

Comment:
Drive by comment:

While this appears to follow the calling convention that LLVM uses, it does _not_ follow the calling convention that avr-gcc uses: https://godbolt.org/z/azq4749hd
`b` is only a single byte on avr-gcc.
(I also see now that the C implementation in compiler-rt has the same bug, but only realized it now).

Reply:
AIUI, these code paths are skipped when compiling for `avr`, precisely because there's a pass which gets rid of these types of shifts in LLVM. So I made no accommodations for its nonstandard calling conventions in this PR.

Reply:
> because there's a pass which gets rid of these types of shifts in LLVM

I know, because I wrote that pass ;)
But it only works for 32-bit values.`__ashldi3` is a 64-bit shift function, which is emitted as a regular compiler-rt call. Proof: https://godbolt.org/z/zqWfE6ndq

Anyway, it doesn't really matter as LLVM uses `i16` and not `i8` like avr-gcc does. This PR would only be buggy for avr-gcc, _right now_ it is safe with LLVM (but this could in theory change as `i8` is more efficient).

Reply:
Minddump for future-me follows:

>b is only a single byte on avr-gcc.

* According to [gcc docs](https://gcc.gnu.org/onlinedocs/gccint/Integer-library-routines.html), `__ashldi3`'s `b` argument is supposed to take an `int`, which must be at least 16 bits.  But indeed, your GCC code snippet is only sending a byte to  `__ashldi3`, while the LLVM version sends 16 bits (the `clr r17` line is clearing the top-most bits).
* On the other hand, `gcc`'s docs also says `__ashldi3` is supposed to return a `long`, and `__ashlti3` is the function that is supposed to return a `long long`. 
* On the other _other hand, `compiler-rt` doesn't seem to support the `t` versions of functions, and tends to hardcode a 64/32-bit int in places gcc's builtins do not<sup>1</sup>. In addition to `compiler-rt` uses `unsigned` in args where gcc's builtins are `signed`.

In other words, it seems like _both_ `gcc` `compiler-rt` builtins tend to do their own things and aren't directly compatible with each other. `avr` using a byte in for the `gcc` `__ashldi3` `b` arg, but 2 bytes for `compiler-rt`'s `__ashldi3` `b` arg is just one example of this. It's not going to be fun :P. Unifying `compiler-rt`'s and `gcc`'s builtins in terms of `compiler_builtins` might be necessary due to the `gcc` backend though.

1. The shift functions' `b` args notwithstanding :P.

Reply:
>But it only works for 32-bit values. __ashldi3 is a 64-bit shift function, which is emitted as a regular compiler-rt call.

How does this not break for Rust code, where all the shift functions are [skipped](https://github.com/rust-lang/compiler-builtins/blob/master/src/int/shift.rs#L72-L124) for AVR (`avr_skip`)?